File: ProgettoScalaChiariMatteo1064144/src/CabinaFotografica.scala
import scala.collection.mutable.Queue
import scala.collection.mutable.ListBuffer


trait getterDati {
  def getCosto: Int
  def getTempoStampa: Int
}

abstract class CartaFotografica extends getterDati {
  val costo:Int
  val tempoStampa:Int
}

class FotoSingola extends CartaFotografica {
  val costo = 1
  val tempoStampa = 2
  def getCosto = this.costo
  def getTempoStampa = this.tempoStampa
}

class QuattroFoto extends CartaFotografica {
  val costo = 3
  val tempoStampa = 6
  def getCosto = this.costo
  def getTempoStampa = this.tempoStampa
}

class SeiFoto extends CartaFotografica {
  val costo = 4
  val tempoStampa = 8
  def getCosto = this.costo
  def getTempoStampa = this.tempoStampa
}

class TimelapseFoto() extends CartaFotografica {
  val costo = 6
  val tempoStampa = 12
  def getCosto = this.costo
  def getTempoStampa = this.tempoStampa
}

class CabinaFototessere {
  val tempoMassimo = 50
    def aggiungiCarta(c: FotoSingola) = CabinaFototessere.carta(0) += c
    def aggiungiCarta(c: QuattroFoto) = CabinaFototessere.carta(1) += c
    def aggiungiCarta(c: SeiFoto) = CabinaFototessere.carta(2) += c
    def aggiungiCarta(c: TimelapseFoto) = CabinaFototessere.carta(3) += c
    
    def stampaSelezione(codice:Int) = {
      if(CabinaFototessere.carta(codice-1).length==0) {
        println("Prodotto Esaurito")
      }
      else {
        var carta = CabinaFototessere.carta(codice-1).dequeue()
        if(CabinaFototessere.credito<carta.getCosto) {
          println("Non hai abbstanza credito, aggiungine altro premendo 7")
          CabinaFototessere.carta(codice-1) += carta
        }
        else{
          aggiungiTempo(carta.getTempoStampa)
          togliCredito(carta.costo)
        println("Erogazione prodotto")
        }
      }
      stampaDisponibilitàCarta
    }
    
    def stampaDisponibilitàCarta = {
      println("Disponibilità")
      println("Singola  4__Foto  6__Foto  Timelapse")
      CabinaFototessere.carta.map(x => print(x.length+"        "))
      println("")
    }
    
    def ricaricaCarta(num:Int)={
      val fotoSingola = new FotoSingola()
      val QuattroFoto = new QuattroFoto()
      val seiFoto = new SeiFoto()
      val timelapseFoto = new TimelapseFoto()
      for(i<- 1 to num){
        aggiungiCarta(fotoSingola)
        aggiungiCarta(QuattroFoto)
        aggiungiCarta(seiFoto)
        aggiungiCarta(timelapseFoto)
      }
    }
    
    def greetings = {
      println("Tempo di Utilizzo: "+getTempo)
      println("Uscita dalla cabina, grazie")
    }
    
    def aggiungiCredito(credito:Int)={
      CabinaFototessere.credito += credito
    }
    
    def togliCredito(credito:Int)={
      CabinaFototessere.credito -= credito
    }
    
    def aggiungiTempo(tempo:Int)={
      CabinaFototessere.tempo += tempo
    }
    
    def stampaCredito = {
      println("Credito: "+CabinaFototessere.credito+" ")
    }
    
    def stampaTempo = {
      println("Tempo: "+CabinaFototessere.tempo+" ")
    }
    
    def inserisciCredito = {
      println("Quanto credito vuoi aggiungere?[Solo numeri]")
      val credito = scala.io.StdIn.readLine()
      var creditoInt:Int = 0
      if(credito!="") creditoInt = credito.toInt
      aggiungiCredito(creditoInt)
    }
    
    def getTempo = {
      CabinaFototessere.tempo
    }
}
object CabinaFototessere {
  private var carta = Array[Queue[CartaFotografica]](Queue[CartaFotografica](), Queue[CartaFotografica](), Queue[CartaFotografica](), Queue[CartaFotografica]())
  private var tempo = 0
  private var credito = 0
}

object test {
   def main(args: Array[String]){
     var cabinaOpen = true
     //println("Ciao Matteo")
     val cabina = new CabinaFototessere()
     cabina.ricaricaCarta(1)
     //cabina.stampaDisponibilitàCarta
     while(cabinaOpen && cabina.getTempo<=cabina.tempoMassimo){
       println("Quale tipo di fortografia desideri? (Inserire il codice relativo)")
       println("1:  Foto Singola")
       println("2:  4 Foto")
       println("3:  6 Foto")
       println("4:  Timelapse Foto")
       println("7:  Inserisci Credito")
       println("8:  Ricarica Carta")
       println("9:  Esci dalla Cabina")
       cabina.stampaDisponibilitàCarta
       cabina.stampaTempo
       cabina.stampaCredito
       val codice = scala.io.StdIn.readLine()
       var codiceInt:Int = 0
       if(codice!="") codiceInt = codice.toInt
       if(codiceInt==8) cabina.ricaricaCarta(1)
       if(codiceInt==7) cabina.inserisciCredito
       else if(codiceInt==9) cabinaOpen = false
       else if(codiceInt<=4 && codiceInt>=1) cabina.stampaSelezione(codiceInt)
     }
     cabina.greetings
   }
}

File: ProgettoC++/src/CabinaFotografica.cpp
#include "CabinaFotografica.h"
#import "CartaFotografica.h"
//#include "CartaUsata.h"
#include "Usata.h"
#include <iostream>
#include <string>

using namespace std;
static bool inizializzato = false;
static CabinaFotografica* cabina;


CabinaFotografica* CabinaFotografica::creaCabinaFotografica(){
	if(!inizializzato){
		cabina = new CabinaFotografica();
		inizializzato = true;
	}
	return cabina;
}
CabinaFotografica::CabinaFotografica(){}


int CabinaFotografica::controllaCredito(CartaFotografica carta){
	if(credito<carta.getPrezzo()){
			cout << "Non hai abbastanza credito, aggiungine altro premendo 7" << endl;
			return 0;
		}
	return 1;
}

void CabinaFotografica::aggiungiTempo(int t){
	tempo+=t;
}
void CabinaFotografica::aggiungiCredito(int c){
	credito+=c;
}
void CabinaFotografica::togliCredito(int c){
	credito=credito-c;
}
void CabinaFotografica::stampaCredito(){
	cout << "Credito: " << credito << endl;
}
void CabinaFotografica::stampaTempo(){
	cout << "Tempo: " << tempo << endl;
}
void CabinaFotografica::stampaInserisciCredito(){
	string inputCreditoTmp;
	int inputCredito = 0;
	cout << "Quanto credito vuoi aggiungere?[Solo numeri] " << tempo << endl;
	cin>>inputCreditoTmp;
	if(inputCreditoTmp!="") inputCredito = stoi(inputCreditoTmp);
	aggiungiCredito(inputCredito);
}

void CabinaFotografica::aggiungiCarta(FotoSingola f){
	fotoSingola.push_back(f);
}
void CabinaFotografica::aggiungiCarta(QuattroFoto f){
	quattroFoto.push_back(f);
}
void CabinaFotografica::aggiungiCarta(SeiFoto f){
	seiFoto.push_back(f);
}
void CabinaFotografica::aggiungiCarta(TimelapseFoto f){
	timelapseFoto.push_back(f);
}
void CabinaFotografica::chiediFiltro(){
	FiltroSeppia fs = FiltroSeppia();
	FiltroBellezza fb = FiltroBellezza();
	string tmpCodice;
	int codice;
	cout << "Quale filtro vuoi applicare" << endl;
	cout << "1:Filtro Seppia - Costo:"<< fs.getPrezzo()<<" - Tempo: "<<fs.getTempo()<<endl;
	cout << "2:Filtro Bellezza - Costo:"<< fb.getPrezzo()<<" - Tempo: "<<fb.getTempo()<<endl;
	cout << "3:Nessun Filtro - Costo: 0 - Tempo: 0"<<endl;

	cin>>tmpCodice;
	if(tmpCodice!="") codice = stoi(tmpCodice);
	//cout << "Codice premuto: " << codice << endl;
	if(codice == 1) {
		aggiungiTempo(fs.getTempo());
		togliCredito(fs.getPrezzo());
	}
	if(codice == 2) {
		aggiungiTempo(fb.getTempo());
		togliCredito(fb.getPrezzo());
	}
	else if(codice == 3) return;
	if(codice<1 || codice >3) return;

}
void CabinaFotografica::buttaCartaUtilizzata(CartaFotografica c){
	CartaUsata cu = CartaUsata(c, tempo) ;
	cartaUsata.push_back(shared_ptr<CartaUsata>(new CartaUsata(c,tempo)));
}
void CabinaFotografica::stampaCartaUsata(){
	for(int i=0; i<cartaUsata.size(); i++){
		cout<<"Carta usata Tempo utilizzo: "<<cartaUsata[i].get()->tempoUtilizzo<<endl;
	}
	if(cartaUsata.size()==0) cout<<"Carta Usata Vuota" <<endl;
}

void CabinaFotografica::svuotaCartaUsata(){
	if(cartaUsata.size()==0) {cout<<"Nessuna carta da svuotare"<<endl;}
	int dim = cartaUsata.size();
	for(int i=0; i<dim; i++){
		cartaUsata.pop_back();
	}
	cout<<"Carta eliminata, dim cartaUsata: "<<cartaUsata.size()<<endl;
}

void CabinaFotografica::stampaSelezione(int codice){
	CartaFotografica carta;
	if(codice==1){
		if(fotoSingola.size()==0){
			cout << "Prodotto Esaurito" << endl; return;
		}
		else {
			carta = fotoSingola.back();
			if(controllaCredito(carta)==1){
				fotoSingola.pop_back();
				chiediFiltro();
				aggTempoTogliCredito(carta.getTempo(), carta.getPrezzo());
				//aggiungiTempo(carta.getTempo());
				//togliCredito(carta.getPrezzo());
				buttaCartaUtilizzata(carta);
				cout << "Erogazione Prodotto" << endl;
			}

		}

	}
	if(codice==2){
			if(quattroFoto.size()==0){
				cout << "Prodotto Esaurito" << endl;
				 return;
			}
			else {
				carta=quattroFoto.back();
				if(controllaCredito(carta)==1){
					quattroFoto.pop_back();
					chiediFiltro();
					aggiungiTempo(carta.getTempo());
					togliCredito(carta.getPrezzo());
					buttaCartaUtilizzata(carta);
					cout << "Erogazione Prodotto" << endl;
				}
			}
		}
	if(codice==3){
			if(seiFoto.size()==0){
				cout << "Prodotto Esaurito" << endl;
				 return;
			}
			else {
				carta=seiFoto.back();
				if(controllaCredito(carta)==1){
					seiFoto.pop_back();
					chiediFiltro();
					aggiungiTempo(carta.getTempo());
					togliCredito(carta.getPrezzo());
					buttaCartaUtilizzata(carta);
					cout << "Erogazione Prodotto" << endl;
				}
			}
		}
	if(codice==4){
			if(timelapseFoto.size()==0) {
				cout << "Prodotto Esaurito" << endl;
				 return;
			}
			else {
				carta=timelapseFoto.back();
				if(controllaCredito(carta)==1){
					timelapseFoto.pop_back();
					chiediFiltro();
					aggiungiTempo(carta.getTempo());
					togliCredito(carta.getPrezzo());
					buttaCartaUtilizzata(carta);
					cout << "Erogazione Prodotto" << endl;
				}
			}
		}

}

void CabinaFotografica::stampaDisponibilitaCarta(){
	cout << "Disponibilità:" << endl;
	cout << "Singola  4__Foto  6__Foto  Timelapse" <<endl;
	cout << fotoSingola.size() <<"        "<< quattroFoto.size() <<"        "<< seiFoto.size() <<"        "<< timelapseFoto.size() << endl;
}
void CabinaFotografica::ricaricaCarta(int n){
	FotoSingola fs = FotoSingola();
	QuattroFoto qf = QuattroFoto();
	SeiFoto sf = SeiFoto();
	TimelapseFoto tf = TimelapseFoto();

	fotoSingola.push_back(fs);
	quattroFoto.push_back(qf);
	seiFoto.push_back(sf);
	timelapseFoto.push_back(tf);
}
void CabinaFotografica::greetings(){
	//cout <<"dim pre svuotamento "<<cartaUsata.size()<<endl;
	cout << "Tempo di utilizzo: " << tempo << endl;
	svuotaCartaUsata();
	//stampaCartaUsata();
	cout << "Uscita dalla cabina, grazie" << endl;
}

CabinaFotografica::~CabinaFotografica(){

	delete cabina;
}

void CabinaFotografica::show(){

	int cabinaOpen = 1;
	int codice;
	string tmpCodice;
	ricaricaCarta(1);
	while(cabinaOpen==1){
		cout << "Quale tipo di fortografia desideri? (Inserire il codice relativo)" << endl;
		cout << "1:  Foto Singola" << endl;
		cout << "2:  4 Foto" << endl;
		cout << "3:  6 Foto" << endl;
		cout << "4:  Timelapse Foto" << endl;
		cout << "6:  Stampa Carta Usata" << endl;
		cout << "7:  Inserisci Credito" << endl;
		cout << "8:  Ricarica Carta" << endl;
		cout << "9:  Esci dalla Cabina" << endl;
		stampaDisponibilitaCarta();
		stampaTempo();
		stampaCredito();
		cin>>tmpCodice;
		if(tmpCodice!="") codice = stoi(tmpCodice);
		//cout << "Codice premuto: " << codice << endl;
		if(codice ==8) ricaricaCarta(1);
		if(codice ==7) stampaInserisciCredito();
		else if(codice==9) cabinaOpen=0;
		else if(codice<=4 && codice>=1) stampaSelezione(codice);
		if(codice==6) stampaCartaUsata();

	};
	greetings();
}
